We have to tell cmake that the openSUSE packager removed the boost::system target.
The headers subpackage provides the necessary Boost:system header file.

diff -ur --no-dereference arrow-apache-arrow-21.0.0.orig/cpp/cmake_modules/ThirdpartyToolchain.cmake arrow-apache-arrow-21.0.0/cpp/cmake_modules/ThirdpartyToolchain.cmake
--- arrow-apache-arrow-21.0.0.orig/cpp/cmake_modules/ThirdpartyToolchain.cmake	2025-07-11 09:44:45.000000000 +0200
+++ arrow-apache-arrow-21.0.0/cpp/cmake_modules/ThirdpartyToolchain.cmake	2025-09-26 20:53:58.409119646 +0200
@@ -1259,7 +1259,7 @@
     set(Boost_USE_STATIC_LIBS ON)
   endif()
   if(ARROW_BOOST_REQUIRE_LIBRARY)
-    set(ARROW_BOOST_COMPONENTS filesystem system)
+    set(ARROW_BOOST_COMPONENTS filesystem)
     if(ARROW_FLIGHT_SQL_ODBC AND MSVC)
       list(APPEND ARROW_BOOST_COMPONENTS locale)
     endif()
diff -ur --no-dereference arrow-apache-arrow-21.0.0.orig/cpp/src/arrow/io/CMakeLists.txt arrow-apache-arrow-21.0.0/cpp/src/arrow/io/CMakeLists.txt
--- arrow-apache-arrow-21.0.0.orig/cpp/src/arrow/io/CMakeLists.txt	2025-07-11 09:44:45.000000000 +0200
+++ arrow-apache-arrow-21.0.0/cpp/src/arrow/io/CMakeLists.txt	2025-09-26 20:53:51.229519926 +0200
@@ -30,7 +30,7 @@
                  EXTRA_LINK_LIBS
                  arrow::hadoop
                  Boost::filesystem
-                 Boost::system)
+                 Boost::headers)
 endif()
 
 add_arrow_test(memory_test PREFIX "arrow-io")
